# Thesis
Repository contains the following:
- Thesis file.
- Code used while working.  
- Posters and presentations presented during the master.

**Thesis details:**




<p align="center">
**Towards realistic sequence simulations for the reconstruction of mega phylogenies** <br/><br/>
Thesis submitted towards the M.Sc. degree in Bioinformatics <br/> 
in Tel-Aviv University <br/> 
by  <br/>
Nomi Hadar <br/> 
 
 
The research was performed in the Department <br/> 
of Molecular Biology and Ecology of Plants <br/> 

Under the supervision of <br/> 
**Prof. Itay Mayrose** <br/> 

June 2018 <br/> 

</p>


**Abstract:**

The reconstruction of large phylogenies encompassing thousands of species enables us a wide view on evolutionary processes and serve as the starting point for the study of numerous evolutionary phenomena. Methods for phylogeny reconstruction are based on an input multiple sequence alignment (MSA), encompassing multiple loci. The extent to which the inferred phylogeny will reflect the true history is thus directly dependent on the accuracy of the alignment. However, the task of sequence alignment becomes more challenging with the amount of divergence among the analyzed sequences. Thus, the reconstruction of large alignments consisting of broadly sampled sequences might lead to poor alignments and inaccurate trees. In this study, I initially aimed to evaluate the effect of alignment complexity on the accuracy of the inferred phylogenies through sequence simulations. Particularly, I examined the hypothesis that the more complex the alignment is the less accurate the inferred phylogeny will be. Such effect was indeed observed; however, it was lower than expected. I suspect that sequence simulations are not realistic enough, rendering them non- informative to study the accuracy of mega phylogenies. My research thus focused on identifying the shortcomings of current simulation approaches and on the pathways by which sequence simulations could be improved. First, I found that available methodologies for inferring the parameters controlling indel dynamics fail to produce alignments that resemble the reference alignments. This required me to develop a new procedure to infer these parameters. Second, using alignment reliability measures, I found that the simulated MSAs differ drastically in their alignment complexity from real MSA, i.e., the complexity of the simulated MSAs was much lower than that of the real one. Third, I found that the inclusion of short, partial, sequences had a major effect on alignment complexity. Fourth, I found that the power-law distribution that is commonly used to describe the length-distribution of indel events, may not be the most suitable one. In summary, my conclusion is that the sequence simulation field is in its early stages, and much study is still required to achive more realistic simulations. Until then, the conclusions that are drawn based on current simulations should be taken with great caution.




